tests/lax_test.py

```diff
@@ -493,6 +493,113 @@ def fun(lhs, rhs):
+           ),
+       "lhs_shape": lhs_shape,
+       "filter_shape": filter_shape,
+       "dtype": dtype,
+       "strides": strides,
+       "padding": padding,
+       "dimension_numbers": dim_nums,
+       "rng_factory": rng_factory,
+       "precision": precision
+      }
+      for dtype in inexact_dtypes
+      for rng_factory in [jtu.rand_small]
+      for lhs_shape, filter_shape, strides, padding, dim_nums in [
+          ((2, 5), (), (), [], ("NC", "OI", "CN")),
+          ((2, 3, 4), (2,), (2,), [(0, 2)], ("CNH", "OHI", "HNC")),
+          ((3, 1, 4, 5), (1, 3), (1, 3), [(3, 1), (2, 2)],
+           ("NCHW", "OIHW", "NCHW")),
+          ((3, 2, 5, 6), (4, 3), (4, 3), [(5, 2), (2, 4)],
+           None),
+          ((1, 2, 3, 4), (1, 1), (1, 1), [(0, 0), (0, 0)],
+           ("NCWH", "OHWI", "CNHW")),
+          ((1, 2, 3, 4), (3, 2), (1, 1), [(0, 0), (0, 0)],
+           ("CWHN", "HOWI", "NCHW")),
+          ((2, 3, 4, 5, 6), (2, 1, 3), (2, 1, 3), [(1, 2), (5, 3), (3, 5)],
+           ("NHWDC", "HDIWO", "DCWNH"))
+      ]
+      for precision in [None,
+                        lax.Precision.DEFAULT,
+                        lax.Precision.HIGH,
+                        lax.Precision.HIGHEST]
+      ))
+  def testConvGeneralDilatedPatches(self, lhs_shape, filter_shape, dtype,
+                                    strides, padding, dimension_numbers,
+                                    rng_factory, precision):
+    rng = rng_factory(self.rng())
+    lhs = rng(lhs_shape, dtype)
+
+    if dimension_numbers is None:
+      lhs_spec, rhs_spec, out_spec = "NCHW", "OIHW", "NCHW"
+    else:
+      lhs_spec, rhs_spec, out_spec = dimension_numbers
+
+    filter_spec = ''.join(c for c in rhs_spec if c not in ('I', 'O'))
+    patches_spec = out_spec.replace('C', 'C' + filter_spec.lower())
+
+    full_padding = []
+    for c in lhs_spec:
+      if c in ('N', 'C'):
+        full_padding += [(0, 0)]
+      else:
+        full_padding += [padding[filter_spec.index(c)]]
+
+    lhs_padded = np.pad(lhs, full_padding, 'constant')
+    out = lax.transpose(lhs_padded, [lhs_spec.index(c) for c in out_spec])
+
+    patches = lax.conv_general_dilated_patches(
+        lhs=lhs,
+        filter_shape=filter_shape,
+        window_strides=strides,
+        padding=padding,
+        dimension_numbers=dimension_numbers,
+        precision=precision
+    )
+
+    source = []
+
+    # Test that output spatial shape is factored into `#patches x patch_size`.
+    for c in out_spec:
+      out_c = out.shape[out_spec.index(c)]
+      patch_c = patches.shape[out_spec.index(c)]
+
+      if c == 'N':
+        self.assertEqual(out_c, patch_c)
+      elif c == 'C':
+        self.assertEqual(out_c * np.prod(filter_shape), patch_c)
+      else:
+        self.assertEqual(out_c, patch_c * filter_shape[filter_spec.index(c)])
+
+        source += [patches_spec.index(c), patches_spec.index(c.lower())]
+
+    # Test that stacking patches together gives the source image, padded.
```

Comment:
I am correct in understanding that this property only holds for carefully chosen pairs of `filter_shape` and `strides`?

It would be nice to have at least one test cases with explicitly constructed inputs where this doesn't hold.

e.g., patches of size 3 in 1D with strides=1:
[1, 2, 3, 4, 5]
-> [[0, 0, 1], [0, 1, 2], [1, 2, 3], [2, 3, 4], [3, 4, 5], [4, 5, 0], [5, 0, 0]]

Reply:
Great suggestion, done!